include/internal/quic_fc.h

```diff
@@ -0,0 +1,250 @@
+     * (STREAM frame offset + payload length) we have seen from a STREAM frame
+     * yet.
+     */
+    uint64_t        cwm, swm, rwm, esrwm, hwm, cur_window_size, max_window_size;
+    OSSL_TIME       epoch_start;
+    OSSL_TIME       (*now)(void *arg);
+    void            *now_arg;
+    QUIC_RXFC       *parent;
+    unsigned char   error_code, has_cwm_changed, is_fin;
+};
+
+/*
+ * Initialises an RX flow controller. conn_rxfc should be non-NULL and point to
+ * a connection-level RXFC if the RXFC is for stream-level flow control, and
+ * NULL otherwise. initial_window_size and max_window_size specify the initial
+ * and absolute maximum window sizes, respectively. Window size values are
+ * expressed in bytes and determine how much credit the RXFC extends to the peer
+ * to transmit more data at a time.
+ */
+int ossl_quic_rxfc_init(QUIC_RXFC *rxfc, QUIC_RXFC *conn_rxfc,
+                        uint64_t initial_window_size,
+                        uint64_t max_window_size,
+                        OSSL_TIME (*now)(void *arg),
+                        void *now_arg);
+
+/*
+ * Gets the parent (i.e., connection-level) RXFC. Returns NULL if called on a
+ * connection-level RXFC.
+ */
+QUIC_RXFC *ossl_quic_rxfc_get_parent(QUIC_RXFC *rxfc);
+
+/*
+ * To be called whenever a STREAM frame is received.
+ *
+ * end is the value (offset + len), where offset is the offset field of the
+ * STREAM frame and len is the length of the STREAM frame's payload in bytes.
+ *
+ * is_fin should be 1 if the STREAM frame had the FIN flag set and 0 otherwise.
+ *
+ * conn_rxfc should point to a connection-level RXFC, which will have its state
+ * updated correctly by the stream-level RXFC.
+ *
+ * This function may be used on a stream-level RXFC only.
+ *
+ * You should check ossl_quic_rxfc_has_error() on both connection-level and
+ * stream-level RXFCs after calling this function, as an incoming STREAM frame
+ * may cause flow control limits to be exceeded by an errant peer. This
+ * function still returns 1 in this case, as this is not a caller error.
+ *
+ * Returns 1 on success or 0 on failure.
+ */
+int ossl_quic_rxfc_on_rx_stream_frame(QUIC_RXFC *rxfc,
+                                      uint64_t end, int is_fin);
+
+/*
+ * To be called whenever controlled bytes are retired, i.e. when bytes are
+ * dequeued from a QUIC stream and passed to the application. num_bytes
+ * is the number of bytes which were passed to the application.
+ *
+ * You should call this only on a stream-level RXFC. Pass a connection-level
+ * RXFC as conn_rxfc; this function will update its state automatically.
+ *
+ * rtt should be the current best understanding of the RTT to the peer, as
+ * offered by the Statistics Manager.
+ *
+ * You should check ossl_quic_rxfc_has_cwm_changed() after calling this
+ * function, as it may have caused the RXFC to decide to grant more flow control
+ * credit to the peer.
```

Comment:
Supposedly the ossl_quic_rxfc_on_retire() is called from the stream receive buffers when the front-end reads data out of it. The ossl_quic_rxfc_has_cwm_changed() is however something that should be called by the connection state machine or whatever is creating the control frames such as the MAX_STREAM_DATA or MAX_DATA. Would it make sense to instead generate an event from the flow control directly?
Another option is that the front end i/o function triggers an event in connection state machine to check for the cwm change.

Reply:
So basically querying this flag vs a callback. Either works, it wouldn't be much work to change it. I figure we can change this later if it makes sense to do so when we consume this module.